sympy/combinatorics/free_group.py

```diff
@@ -0,0 +1,938 @@
+
+    References
+    ==========
+
+    [1] http://www.gap-system.org/Manuals/doc/ref/chap37.html
+
+    [2] https://en.wikipedia.org/wiki/Free_group
+
+    See Also
+    ========
+
+    sympy.polys.rings.PolyRing
+
+    """
+    is_associative = True
+    is_group = True
+    is_FreeGroup = True
+    is_PermutationGroup = False
+
+    def __new__(cls, symbols):
+        symbols = tuple(_parse_symbols(symbols))
+        rank = len(symbols)
+        _hash = hash((cls.__name__, symbols, rank))
+        obj = _free_group_cache.get(_hash)
+
+        if obj is None:
+            obj = object.__new__(cls)
+            obj._hash = _hash
+            obj._rank = rank
+            # dtype method is used to create new instances of FreeGroupElm
+            obj.dtype = type("FreeGroupElm", (FreeGroupElm,), {"group": obj})
+            obj.symbols = symbols
+            obj.generators = obj._generators()
+            obj._gens_set = set(obj.generators)
+            for symbol, generator in zip(obj.symbols, obj.generators):
+                if isinstance(symbol, Symbol):
+                    name = symbol.name
+                    if hasattr(obj, name):
+                        setattr(obj, name, generator)
+
+            _free_group_cache[_hash] = obj
+
+        return obj
+
+    def _generators(group):
+        """Returns the generators of the FreeGroup
+
+        Examples
+        ========
+
+        >>> from sympy.combinatorics.free_group import free_group
+        >>> F, x, y, z = free_group("x, y, z")
+        >>> F.generators
+        (x, y, z)
+
+        """
+        gens = []
+        for sym in group.symbols:
+            elm = ((sym, 1),)
+            gens.append(group.dtype(elm))
+        return tuple(gens)
+
+    def __getitem__(self, i):
+        return self.generators[i]
+
+    def clone(self, symbols=None):
+        return self.__class__(symbols or self.symbols)
+
+    def __contains__(self, i):
+        """Return True if `i` is contained in FreeGroup.
+        """
+        if not isinstance(i, FreeGroupElm):
+            raise TypeError("FreeGroup contains only FreeGroupElm as elements "
+                        ", not elements of type %s" % type(i))
+        group = i.group
+        return self == group
+
+    def __hash__(self):
+        return self._hash
+
+    def __len__(self):
+        return self.rank
+
+    def __str__(self):
+        if self.rank > 30:
+            str_form = "<free group with %s generators>" % self.rank
+        else:
+            str_form = "<free group on the generators "
+            gens = self.generators
+            str_form += str(gens) + ">"
+        return str_form
+
+    __repr__ = __str__
+
+    def __getitem__(self, index):
+        symbols = self.symbols[index]
+        return self.clone(symbols=symbols)
+
+    def __eq__(self, other):
+        """No ``FreeGroup`` is equal to any "other" ``FreeGroup``.
+        """
+        return self is other
```

Comment:
This could be changed. Its not something under a priority, so I left it as it is.